authentication.py

```diff
@@ -0,0 +1,84 @@
+import urllib
+import json
+import endpoints
+import os
+import logging
+
+from google.appengine.api import memcache
+from google.appengine.api import urlfetch
+
+_ACCEPTED_AUDIENCE_PREFIXES = os.environ['BOND_ACCEPTED_AUDIENCE_PREFIXES'].split()
+_ACCEPTED_EMAIL_SUFFIXES = os.environ['BOND_ACCEPTED_EMAIL_SUFFIXES'].split()
+_TOKENINFO_URL = 'https://www.googleapis.com/oauth2/v1/tokeninfo'
+
+
+class UserInfo:
+    def __init__(self, id, email, token, expires_in):
+        self.id = id
+        self.email = email
+        self.token = token
+        self.expires_in = expires_in
+
+    def __str__(self):
+        return 'id: {}, email: {}, token: {}, expires_id: {}'.format(self.id, self.email, self.token, self.expires_in)
+
+
+def _token_info(token):
+    result = urlfetch.fetch(
+        '%s?%s' % (_TOKENINFO_URL, urllib.urlencode({'access_token': token})))
+    if result.status_code != 200:
+        raise endpoints.InternalServerErrorException(message='Token info endpoint returned status {}: {}'.format(result.status_code, result.content))
+    return result
+
+
+def require_user_info(request_state, token_info_fn=_token_info):
+    """Get the user's info from cache or from google if not in cache, throwing unauthorized errors as appropriate
+
+    :param request_state: self.request_state from a class extending protorpc.remote.Service
+    :param token_info_fn: function to get token info, defaults to google http request, override for testing
+    :return: UserInfo instance
+    """
+    auth_header = request_state.headers.get('Authorization')
+    if auth_header is None:
+        raise endpoints.UnauthorizedException(message='Request missing Authorization header.')
+
+    auth_header_parts = auth_header.split()
+    if len(auth_header_parts) != 2 or auth_header_parts[0].lower() != 'bearer':
+        raise endpoints.UnauthorizedException(message='Malformed Authorization header.')
+
+    token = auth_header_parts[1]
+
+    user_info = memcache.get(key=token)
+    if user_info is None:
+        logging.debug("auth token cache miss for token %s", token)
+        user_info = _fetch_user_info(token, token_info_fn)
+        # cache for 10 minutes or until token expires
+        memcache.add(key=token, value=user_info, time=min([user_info.expires_in, 600]))
```

Comment:
`min([None, 600]) == None` so if `user_info.expires_in` is ever blank for some reason, this might be a problem